File: Be-siukat-new/siukat-service-dev_snmptn_2022/services/pendukung.js
var Pendukung = require('../models/pendukung');
var LogPendukung = require('../models/log_pendukung');

var pendukung = function () {};

pendukung.prototype.add = function(req, atribut){
  return new Promise(function(resolve, reject){
    Pendukung.create({
      no_peserta: req.no_peserta,
      tanggungan: req.tanggungan,
      scan_pernyataan_ukt_tinggi: req.scan_pernyataan_ukt_tinggi,
      scan_pernyataan_kebenaran: req.scan_pernyataan_kebenaran,
      scan_kk: req.scan_kk,
      atribut: atribut
    }, {
      fields: [
        'no_peserta',
        'tanggungan',
        'scan_pernyataan_ukt_tinggi',
        'scan_pernyataan_kebenaran',
        'scan_kk',
        'atribut'
      ]
    }).then(function(response){
      resolve(response)
    }).catch(function(err){
      reject(err)
    })
  })
}

pendukung.prototype.edit = function(req, no_peserta, atribut){
  return new Promise(function(resolve, reject){
    var data = {
      tanggungan: req.body.tanggungan,
    }

    if(req.files["file_scan_pernyataan_ukt_tinggi"] !== undefined){
      data.scan_pernyataan_ukt_tinggi = req.files["file_scan_pernyataan_ukt_tinggi"][0].filename
    }
    if(req.files["file_scan_pernyataan_kebenaran"] !== undefined){
      data.scan_pernyataan_kebenaran = req.files["file_scan_pernyataan_kebenaran"][0].filename
    }
    if(req.files["file_scan_kk"] !== undefined){
      data.scan_kk = req.files["file_scan_kk"][0].filename
    }

    Pendukung.update(data, {
      where: {
        no_peserta: no_peserta,
		atribut: atribut
      }
    }).then(function(response){
      resolve(response)
    }).catch(function(err){
      reject(err)
    })
  })
}

pendukung.prototype.addLog = function(users, atribut, executor, timestamp){
  return new Promise(function(resolve, reject){
    LogPendukung.create({
      no_peserta: users.no_peserta,
      tanggungan: users.tanggungan,
      scan_pernyataan_ukt_tinggi: users.scan_pernyataan_ukt_tinggi,
      scan_pernyataan_kebenaran: users.scan_pernyataan_kebenaran,
      scan_kk: users.scan_kk,
      atribut: atribut,
      executor: executor,
      timestamp: timestamp
    }, {
      field: [
        'no_peserta',
        'tanggungan',
        'scan_pernyataan_ukt_tinggi',
        'scan_pernyataan_kebenaran',
        'scan_kk',
        'atribut',
        'executor',
        'timestamp'
      ]
    }).then(function(response){
      resolve(response)
    }).catch(function(err){
      reject(err)
    })
  })
}

pendukung.prototype.getByLoggedIn = function(no_peserta){
  return new Promise(function(resolve, reject){
    Pendukung.findOne({
      where: {
        no_peserta: no_peserta
      }
    }).then(function(response){
      resolve(response)
    }).catch(function(err){
      reject(err)
    })
  })
}

pendukung.prototype.checkData = function(no_peserta, users){
  return new Promise((resolve, reject) => {
    Pendukung.findOne({
      where: {
        no_peserta: no_peserta
      },
      raw: true,
    }).then(function (response){
      var result = true
      var data = {}
      if(users.ukt_tinggi === "ya"){
        data.scan_pernyataan_ukt_tinggi = response.scan_pernyataan_ukt_tinggi
      }else{
        delete response.scan_pernyataan_ukt_tinggi
        delete response.scan_pernyataan_kebenaran
        data = response
      }
      for(var key in data){
        result *= (data[key] !== 0 && data[key] !== "")
      }
      resolve(result)
    }).catch(function (err) {
      reject(err)
    })
  })
}

pendukung.prototype.checkDataVerifikasi = function(no_peserta, users){
  return new Promise((resolve, reject) => {
    Pendukung.findOne({
      where: {
        no_peserta: no_peserta
      },
      raw: true,
    }).then(function (response){
      var result = true
      var data = {}
      if(users.ukt_tinggi === "ya"){
        data.scan_pernyataan_ukt_tinggi = response.scan_pernyataan_ukt_tinggi
      }else{
        data.scan_pernyataan_kebenaran = response.scan_pernyataan_kebenaran
      }
      for(var key in data){
        result *= (data[key] !== 0 && data[key] !== "")
      }
      resolve(result)
    }).catch(function (err) {
      reject(err)
    })
  })
}

module.exports = pendukung

File: BackEnd-Siukat/services/pendukung.go
package services

import (
	"BackEnd-Siukat/config"
	"BackEnd-Siukat/models"
	"time"
)

type PendukungService struct{}

func (s *PendukungService) Add(req models.Pendukung, atribut string) (models.Pendukung, error) {
	req.Atribut = atribut
	err := config.DB.Create(&req).Error
	return req, err
}

func (s *PendukungService) Edit(req models.Pendukung, noPeserta string, existing models.Pendukung, atribut string) (models.Pendukung, error) {
	data := map[string]interface{}{
		"tanggungan": req.Tanggungan,
	}

	if req.ScanPernyataanUktTinggi != "" {
		data["scan_pernyataan_ukt_tinggi"] = req.ScanPernyataanUktTinggi
	} else {
		data["scan_pernyataan_ukt_tinggi"] = existing.ScanPernyataanUktTinggi
	}

	if req.ScanPernyataanKebenaran != "" {
		data["scan_pernyataan_kebenaran"] = req.ScanPernyataanKebenaran
	} else {
		data["scan_pernyataan_kebenaran"] = existing.ScanPernyataanKebenaran
	}

	if req.ScanKk != "" {
		data["scan_kk"] = req.ScanKk
	} else {
		data["scan_kk"] = existing.ScanKk
	}

	if err := config.DB.Model(&models.Pendukung{}).Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).Updates(data).Error; err != nil {
		return models.Pendukung{}, err
	}
	var res models.Pendukung
	config.DB.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&res)
	return res, nil
}

func (s *PendukungService) AddLog(user models.Pendukung, atribut string, executor string, timestamp *time.Time) error {
	logLst := models.LogPendukung{
		NoPeserta:               user.NoPeserta,
		Tanggungan:              user.Tanggungan,
		ScanPernyataanUktTinggi: user.ScanPernyataanUktTinggi,
		ScanPernyataanKebenaran: user.ScanPernyataanKebenaran,
		ScanKk:                  user.ScanKk,
		Atribut:                 atribut,
		Executor:                executor,
		Timestamp:               timestamp,
	}
	return config.DB.Create(&logLst).Error
}

func (s *PendukungService) GetByLoggedIn(noPeserta string) (models.Pendukung, error) {
	var res models.Pendukung
	err := config.DB.Where("no_peserta = ?", noPeserta).First(&res).Error
	return res, err
}

func (s *PendukungService) CheckData(noPeserta string, uktTinggi string) (bool, error) {
	var res models.Pendukung
	if err := config.DB.Where("no_peserta = ?", noPeserta).First(&res).Error; err != nil {
		return false, err
	}

	if uktTinggi == "ya" {
		if res.ScanPernyataanUktTinggi == "" {
			return false, nil
		}
	} else {
		if res.Tanggungan == 0 || res.ScanKk == "" {
			return false, nil
		}
	}
	return true, nil
}

func (s *PendukungService) CheckDataVerifikasi(noPeserta string, uktTinggi string) (bool, error) {
	var res models.Pendukung
	if err := config.DB.Where("no_peserta = ?", noPeserta).First(&res).Error; err != nil {
		return false, err
	}

	if uktTinggi == "ya" {
		if res.ScanPernyataanUktTinggi == "" {
			return false, nil
		}
	} else {
		if res.ScanPernyataanKebenaran == "" {
			return false, nil
		}
	}
	return true, nil
}
